### PR TITLE
HIVE-26149: Non blocking DROP DATABASE implementation

### PR DESCRIPTION
**Pull request by deniskuzZ**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/database/drop/DropDatabaseAnalyzer.java [hunk 1]
```diff
@@ -49,28 +52,37 @@ public void analyzeInternal(ASTNode root) throws SemanticException {
...
-        for (Table table : db.getAllTableObjects(databaseName)) {
+        List<Table> tables = db.getAllTableObjects(databaseName);
+        allTablesWithSuffix = tables.stream().allMatch(
+            table -> AcidUtils.isTableSoftDeleteEnabled(table, conf));
+        for (Table table : tables) {
           // We want no lock here, as the database lock will cover the tables,
           // and putting a lock will actually cause us to deadlock on ourselves.
-          outputs.add(new WriteEntity(table, WriteEntity.WriteType.DDL_NO_LOCK));
+          outputs.add(
+            new WriteEntity(table, isSoftDelete && !allTablesWithSuffix ?
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/database/drop/DropDatabaseAnalyzer.java:72]: Nit: Could we create boolean variables with descriptive names? It is hard to follow what happens here.

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java [hunk 2]
```diff
@@ -661,16 +662,36 @@ public void dropDatabase(String name, boolean deleteData, boolean ignoreUnknownD
    */
   public void dropDatabase(String name, boolean deleteData, boolean ignoreUnknownDb, boolean cascade)
       throws HiveException, NoSuchObjectException {
+    dropDatabase(
+      new DropDatabaseDesc(name, ignoreUnknownDb, cascade, deleteData));
+  }
+
+  public void dropDatabase(DropDatabaseDesc desc)
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java:669]: Nit: I would guess that we do not need the new line here

File: ql/src/test/org/apache/hadoop/hive/ql/TestTxnCommands.java [hunk 3]
```diff
@@ -1685,7 +1688,89 @@ public void testDropWithBaseMultiplePartitions() throws Exception {
...
+  @Test
+  public void testDropDatabaseCascadePerTableNonBlocking() throws Exception {
+    MetastoreConf.setLongVar(hiveConf, MetastoreConf.ConfVars.BATCH_RETRIEVE_MAX, 1);
+    dropDatabaseCascadeNonBlocking();
+  }
+  @Test
+  public void testDropDatabaseCascadePerDbNonBlocking() throws Exception {
+    dropDatabaseCascadeNonBlocking();
+  }
+  private void dropDatabaseCascadeNonBlocking() throws Exception {
```
Comment (pvary) [inline ql/src/test/org/apache/hadoop/hive/ql/TestTxnCommands.java:1701]: Nit: newline

File: ql/src/test/org/apache/hadoop/hive/ql/TestTxnCommands.java [hunk 4]
```diff
@@ -1685,7 +1688,89 @@ public void testDropWithBaseMultiplePartitions() throws Exception {
       }
     }
   }
+  
+  @Test
+  public void testDropDatabaseCascadePerTableNonBlocking() throws Exception {
+    MetastoreConf.setLongVar(hiveConf, MetastoreConf.ConfVars.BATCH_RETRIEVE_MAX, 1);
+    dropDatabaseCascadeNonBlocking();
+  }
+  @Test
```
Comment (pvary) [inline ql/src/test/org/apache/hadoop/hive/ql/TestTxnCommands.java:1697]: Nit: newline

File: ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java [hunk 5]
```diff
@@ -3914,4 +3914,72 @@ private void testRenamePartition(boolean blocking) throws Exception {
     driver.getFetchTask().fetch(res);
     Assert.assertEquals("Expecting 1 rows and found " + res.size(), 1, res.size());
   }
+
+  @Test
+  public void testDropDatabaseNonBlocking() throws Exception {
+    dropDatabaseNonBlocking(false, false);
+  }
+  @Test
```
Comment (pvary) [inline ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java:3922]: nit: newlines

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java [hunk 6]
```diff
@@ -1457,39 +1458,42 @@ public void dropDatabase(String name)
 
   @Override
   public void dropDatabase(String name, boolean deleteData, boolean ignoreUnknownDb)
-      throws NoSuchObjectException, InvalidOperationException, MetaException, TException {
+      throws TException {
```
Comment (pvary) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:1461]: Could we make the old methods deprecated?

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java [hunk 7]
```diff
@@ -1534,43 +1538,50 @@ public void dropDatabase(String catalogName, String dbName, boolean deleteData,
...
+   * @param req
    * @param tableList
-   * @param deleteData
    * @throws TException
    */
-  private void dropDatabaseCascadePerDb(String catName, String dbName, List<String> tableList,
-                                        boolean deleteData) throws TException {
-    String dbNameWithCatalog = prependCatalogToDbName(catName, dbName, conf);
-    List<Table> tables = getTableObjectsByName(catName, dbName, tableList);
+  private void dropDatabaseCascadePerDb(DropDatabaseRequest req, List<String> tableList) throws TException {
```
[the remark this answers is not in the document]
Comment (deniskuzZ) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:1583]: if DB has a mix of soft-delete(prefixed) and managed tables, we acquire exclusive locks and remove as usual managed/external tables, however, for soft-delete tables we acquire excl_write lock and delegate cleanup to the cleaner process.  Read locks are only removed on soft-delete tables.  
Note: if lockless reads are enabled we do not remove the db folder.

[the remark this answers is not in the document]
Comment (pvary) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:1583]: How does this work together with:
```
          // We want no lock here, as the database lock will cover the tables,
          // and putting a lock will actually cause us to deadlock on ourselves.
```

Wouldn't it cause issues with the locks?

[the remark this answers is not in the document]
Comment (deniskuzZ) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:1583]: that remains if soft delete is disabled, otherwise, we'll be locking just tables/not DB with an appropriate type of lock.  
allTablesWithSuffix is an optimization, if all tables under DB are soft-delete eligible - grad just table-level lock.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/database/drop/DropDatabaseAnalyzer.java [hunk 8]
```diff
@@ -49,28 +52,36 @@ public void analyzeInternal(ASTNode root) throws SemanticException {
...
+        List<Table> tables = db.getAllTableObjects(databaseName);
+        allTablesWithSuffix = tables.stream().allMatch(
+            table -> AcidUtils.isTableSoftDeleteEnabled(table, conf));
+        for (Table table : tables) {
+          // Optimization used to limit number of requested locks. Check if table lock is needed or we could get away with single DB level lock,
+          boolean isTableLockNeeded = isSoftDelete && !allTablesWithSuffix;
+          outputs.add(new WriteEntity(table, isTableLockNeeded ?
+            AcidUtils.isTableSoftDeleteEnabled(table, conf) ?
+                WriteEntity.WriteType.DDL_EXCL_WRITE : WriteEntity.WriteType.DDL_EXCLUSIVE :
+            WriteEntity.WriteType.DDL_NO_LOCK));
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/database/drop/DropDatabaseAnalyzer.java:74]: Would this be better:
```
LockType lockType = WriteEntity.WriteType.DDL_NO_LOCK;
if (isTableLockNeeded) {
   lockType = AcidUtils.isTableSoftDeleteEnabled(table, conf) ?
                WriteEntity.WriteType.DDL_EXCL_WRITE : WriteEntity.WriteType.DDL_EXCLUSIVE;
}
outputs.add(new WriteEntity(table, lockType));
```

I think having too many `:` and `?` is really hard to read.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/IMetaStoreClient.java [hunk 9]
```diff
@@ -1869,8 +1880,7 @@ void dropDatabase(String catName, String dbName, boolean deleteData, boolean ign
    * @throws MetaException something went wrong, usually either in the RDBMS or storage.
    * @throws TException general thrift error.
    */
-  default void dropDatabase(String catName, String dbName, boolean deleteData,
-                            boolean ignoreUnknownDb)
+  default void dropDatabase(String catName, String dbName, boolean deleteData, boolean ignoreUnknownDb)
```
Comment (pvary) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/IMetaStoreClient.java:1883]: Nit: I think this could be deprecated as well

**Review by pvary — APPROVED** [review]
Only minor comments left.
Thanks for the fixed @deniskuzZ!

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/database/drop/DropDatabaseAnalyzer.java (hunk 8 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/database/drop/DropDatabaseAnalyzer.java]: refactored
